### PR TITLE
Qcd

### PR DESCRIPTION
**Pull request by valeriadamante**

### CODE REVIEW

File: Analysis/Producer.py [hunk 1]
```diff
@@ -42,15 +50,13 @@ def __init__(self, df):
...
         self.defineAllP4()
         self.createInvMass()
+        self.defineRegions()
 
 
     def skimAnatuple(self):
-        for tau_idx in [1,2]:
-            self.df = self.df.Filter(f'tau{tau_idx}_idDeepTau{self.deepTauYear}{self.deepTauVersion}VSjet >= {Utilities.WorkingPointsTauVSjet.Medium.value}')
-        self.df = self.df.Filter('tau1_charge * tau2_charge <0')
+        self.df = self.df.Filter(f'tau1_idDeepTau{self.deepTauYear}{self.deepTauVersion}VSjet >= {Utilities.WorkingPointsTauVSjet.Medium.value}')
```
Comment (kandrosov) [inline Analysis/Producer.py:57]:
```suggestion
```

File: Analysis/Producer.py [hunk 2]
```diff
@@ -63,7 +69,22 @@ def defineWeights(df_dict):
...
+    hist_data_3 = df_data.Filter('region_C').Histo1D(model, var, "weight").Clone()
+    data_4 = df_data.Filter('region_D').Sum("weight").GetValue()
+    kappa = data_2/data_4
+    for sample in df_dict.keys():
+        if sample=='data' or sample in signals:
+            continue
+        hist_3 = df_dict[sample].Filter("region_C").Histo1D(model,var,"weight").Clone() # OS non-iso
+        hist_data_3.Add(hist_3, -1)
+    hist_data_3.Scale(kappa)
+    return hist_data_3
```
Comment (kandrosov) [inline Analysis/Producer.py:85]: only store the ingredients. do the actual computation in the final loop where `GetValue` is called for all histograms

File: Analysis/Producer.py [hunk 3]
```diff
@@ -63,7 +69,22 @@ def defineWeights(df_dict):
...
         df_dict[sample]=df_dict[sample].Define("weight",weight_str)
 
-def createHistograms(df_dict, var):
+
+def Estimate_QCD(var, df_dict, hist_dict, model, deepTauVersion):
+    df_data = df_dict['data']
+    data_2 = df_data.Filter('region_B').Sum("weight").GetValue()
+    hist_data_3 = df_data.Filter('region_C').Histo1D(model, var, "weight").Clone()
+    data_4 = df_data.Filter('region_D').Sum("weight").GetValue()
+    kappa = data_2/data_4
```
Comment (kandrosov) [inline Analysis/Producer.py:78]: wrong - MC is not subtracted

File: config/plot/2018.yaml [hunk 4]
```diff
@@ -5,3 +5,10 @@ lumi_text:
   text_size: 0.035
   font: 42
   align: right_bottom
+
+cat_text:
+  text: "inclusive"
+  pos_ref: scope_text
+  pos: [0, -0.1 ]
+  text_size: 0.03
+  font: 62
```
Comment (kandrosov) [inline config/plot/2018.yaml:14]:
```suggestion
```
we will have multiple categories for the same year

**Review by kandrosov — CHANGES_REQUESTED** [review]

File: Analysis/Producer.py [hunk 5]
```diff
@@ -30,27 +30,35 @@ def defineAllP4(self):
...
         self.defineAllP4()
         self.createInvMass()
+        self.defineRegions()
 
 
     def skimAnatuple(self):
-        for tau_idx in [1,2]:
-            self.df = self.df.Filter(f'tau{tau_idx}_idDeepTau{self.deepTauYear}{self.deepTauVersion}VSjet >= {Utilities.WorkingPointsTauVSjet.Medium.value}')
-        self.df = self.df.Filter('tau1_charge * tau2_charge <0')
+        self.df = self.df.Filter(f'tau1_idDeepTau{self.deepTauYear()}{self.deepTauVersion}VSjet >= {Utilities.WorkingPointsTauVSjet.Medium.value}')
```
Comment (kandrosov) [inline Analysis/Producer.py:59]:
```suggestion
```
this is should be applied at the baseline level. isn't it?

File: Analysis/Producer.py [hunk 6]
```diff
@@ -117,19 +150,27 @@ def createHistograms(df_dict, var):
             anaskimmer.df = anaskimmer.df.Filter(f"X_mass=={args.mass}")
         anaskimmer.skimAnatuple()
         dataframes[sample] = anaskimmer.df
-
     defineWeights(dataframes)
+
     all_histograms = {}
+    all_sums = {}
```
Comment (kandrosov) [inline Analysis/Producer.py:156]:
```suggestion
```

File: Analysis/Producer.py [hunk 7]
```diff
@@ -117,19 +150,27 @@ def createHistograms(df_dict, var):
...
     vars = args.vars.split(',')
+    all_sums = createSums(dataframes)
     for var in vars:
         hists = createHistograms(dataframes, var)
         all_histograms[var] = hists
 
+    hists_to_plot = {}
     for var in vars:
```
Comment (kandrosov) [inline Analysis/Producer.py:164]: To simplify `GetValue` procedure define:
```python
def GetValues(collection):
    for key, value in collection.items():
        if isinstance(value, dict):
            GetValues(value)
        else:
            collection[key] = value.GetValue()
```
and then call it before entering in the loops:
```
GetValues(all_histograms)
GetValues(all_sums)
```

**Review by kandrosov — CHANGES_REQUESTED** [review]

File: Analysis/Producer.py (hunk 5 above)
Comment (valeriadamante) [reply Analysis/Producer.py:59]: I am not sure. I put it to be safe

File: Analysis/Producer.py [hunk 8]
```diff
@@ -63,6 +71,78 @@ def defineWeights(df_dict):
...
+
+def FixNegativeContributions(histogram, debug_info, negative_bins_info):
+    correction_factor = 0.0000001
+
+    ss_debug = ""
+
+    original_Integral = histogram.Integral(0, histogram.GetNbinsX()+1, "width")
+    ss_debug += "\nSubtracted hist for '{}'.\n".format(histogram.GetName())
+    ss_debug += "Integral after bkg subtraction: {}.\n".format(original_Integral)
+    debug_info = ss_debug
```
Comment (kandrosov) [inline Analysis/Producer.py:88]:
```suggestion
```

File: Analysis/Producer.py [hunk 9]
```diff
@@ -63,6 +71,78 @@ def defineWeights(df_dict):
...
+    if original_Integral < 0:
+        print(debug_info)
+        print("Integral after bkg subtraction is negative for histogram '{}'".format(histogram.GetName()))
+        return False
+
+    ss_negative = ""
```
Comment (kandrosov) [inline Analysis/Producer.py:94]:
```suggestion
```

File: Analysis/Producer.py [hunk 10]
```diff
@@ -63,6 +71,78 @@ def defineWeights(df_dict):
...
+            histogram.GetBinLowEdge(n), histogram.GetBinLowEdge(n+1))
+
+        error = correction_factor - histogram.GetBinContent(n)
+        new_error = math.sqrt(math.pow(error, 2) + math.pow(histogram.GetBinError(n), 2))
+        histogram.SetBinContent(n, correction_factor)
+        histogram.SetBinError(n, new_error)
+
+    RenormalizeHistogram(histogram, original_Integral, True)
+    negative_bins_info = ss_negative
+    return True
```
Comment (kandrosov) [inline Analysis/Producer.py:112]:
```suggestion
    return True, ss_debug, ss_negative
```

**Review by kandrosov — APPROVED** [review]

**Pull request merged**